pkg/kubelet/pluginmanager/pluginwatcher/monitor.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginwatcher
+
+import (
+	"context"
+	"os"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"k8s.io/klog/v2"
+	registerapi "k8s.io/kubelet/pkg/apis/pluginregistration/v1"
+	"k8s.io/kubernetes/pkg/kubelet/pluginmanager/cache"
+	"k8s.io/kubernetes/pkg/kubelet/util"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+const (
+	// dialTimeout is the timeout duration for dialing a plugin socket.
+	dialTimeout time.Duration = time.Second
+
+	// getInfoTimeout is the timeout duration for GetInfo call.
+	getInfoTimeout time.Duration = time.Second
+
+	// pluginListInterval is the duration between getting list of registered plugins from ASW.
+	pluginListInterval time.Duration = time.Second
+
+	// retryInterval is the interval duration between connection checks.
+	connectionCheckInterval time.Duration = time.Second
+
+	// maxFailures is the maximum number of allowed connection monitor failures.
+	// If the number of consecutive failures exceeds this value, the connection is considered dead.
+	maxFailures uint = 2
+)
+
+// pluginConnectionMonitor is a struct to monitor GRPC connections between Kubelet and plugins
+type pluginConnectionMonitor struct {
+	// Directory to watch for plugin sockets
+	sockDir string
+
+	// A sync.Map to store monitored connections
+	connMap sync.Map
+
+	// A channel to stop the monitor
+	stopCh chan struct{}
+	once   sync.Once
+
+	dsw cache.DesiredStateOfWorld
+	asw cache.ActualStateOfWorld
+}
+
+// newPluginConnectionMonitor creates a new plugin connection monitor.
+func newPluginConnectionMonitor(sockDir string, dsw cache.DesiredStateOfWorld, asw cache.ActualStateOfWorld) *pluginConnectionMonitor {
+	return &pluginConnectionMonitor{
+		sockDir: sockDir,
+		connMap: sync.Map{},
+		stopCh:  make(chan struct{}),
+		dsw:     dsw,
+		asw:     asw,
+	}
+}
+
+// start starts the plugin connection monitor.
+func (m *pluginConnectionMonitor) start() {
```

Comment:
with the existing of this function, can we trust that if `GetInfo` successful, we do not need to monitor the health of an actual plugin socket connection? Or for each plugin there will be two of those health loops?

Reply:
This function periodically discovers sockets that are not monitored and starts monitoring go-routine `monitorPluginConnection` for them one go-routine per connection, so there is only one loop per socket/connection.